Pipeline for policy generation from DDS discovery data

ruffsl opened the pull request:
Related: https://github.com/ros2/sros2/issues/112

ruffsl committed:
Add initial template to demangle DDS to SROS2

Signed-off-by: ruffsl <roxfoxpox@gmail.com>

ruffsl committed:
Deduplicate and alpha sort elements/atrabutes

Signed-off-by: ruffsl <roxfoxpox@gmail.com>

ruffsl committed:
Remove dead code

Signed-off-by: ruffsl <roxfoxpox@gmail.com>

ruffsl committed:
Set unique mode for substation

Signed-off-by: ruffsl <roxfoxpox@gmail.com>

ruffsl committed:
Adding deduplicate namespace and compress for profiles
apply template on each profile to avoided crosstalk

Signed-off-by: ruffsl <roxfoxpox@gmail.com>

ruffsl committed:
Check in initial discovery sqlite script

Signed-off-by: ruffsl <roxfoxpox@gmail.com>

mikaelarguedas commented on sros2/sros2/policy/templates/dds/policy.xsl:
```diff
@@ -0,0 +1,494 @@
+    <mapping type="status"   prefix="rt"  suffix="/_action/status"/>
+  </mappings>
+  <permissions>
+    <permission kind="publish"   type="feedback" rule="excute"/>
```
```suggestion
    <permission kind="publish"   type="feedback" rule="execute"/>
```

mikaelarguedas commented on sros2/scripts/dds_sql_to_sros2_policy.py:
```diff
@@ -0,0 +1,172 @@
+# In[ ]:
+
+
+polciy = etree.Element('polciy')
```
```suggestion
policy = etree.Element('policy')
```

mikaelarguedas replied:
Same multiple times below

mikaelarguedas commented on sros2/scripts/dds_sql_to_sros2_policy.py:
```diff
@@ -0,0 +1,172 @@
+        for mode in _df.index.get_level_values('mode').unique():
+            topics = etree.SubElement(profile, 'dds_topics')
+            topics.set(mode, "ALLOW")
+            for dds_topic in df['dds_topic'].loc[namespace, name, :]:
```
I believe this should check for `mode` to not add all topics corresponding to this namespace + name

mikaelarguedas reviewed (commented):
Some things that came up when spot comparing the generated files with the ones from https://github.com/mikaelarguedas/tb3_demo/blob/857e1c62ba06476cda4ba359fdf59d31ba488f91/policies/sqlite_playground/extract_policy_from_discovery_db.py

ruffsl committed:
Fix typo for execute

Signed-off-by: ruffsl <roxfoxpox@gmail.com>

ruffsl committed:
Fix typo for policy

Signed-off-by: ruffsl <roxfoxpox@gmail.com>

ruffsl committed:
Correct indexing to include mode
To avoid provisioning topic to all modes
instead of the topic's respective mode as intended

Signed-off-by: ruffsl <roxfoxpox@gmail.com>

ruffsl committed:
Rename to dds/demangle.xml

Signed-off-by: ruffsl <roxfoxpox@gmail.com>

ruffsl committed:
Clean up notebook into runnable script
Borrowing cli petern from @mikaelarguedas

Signed-off-by: ruffsl <roxfoxpox@gmail.com>

ruffsl commented:
How's it now? See example output:
https://pastebin.com/JhCvvYVh

mikaelarguedas commented on sros2/scripts/dds_sql_to_sros2_policy.py:
```diff
@@ -0,0 +1,143 @@
+    sros2_policy = dds_policy_to_sros2_policy(dds_policy)
+
+    if args.output_policy is not None:
+        with open(args.policy_file, 'wb') as f:
```
```suggestion
        with open(args.output_policy, 'wb') as f:
```

ruffsl replied:
297b049

mikaelarguedas commented on sros2/scripts/dds_sql_to_sros2_policy.py:
```diff
@@ -0,0 +1,143 @@
+
+import argparse
+import os
+import pandas as pd
```
Note for later: needed to install this dep by hand, we should list it in the `package.xml` ([python3-pandas](https://github.com/ros/rosdistro/blob/a63054565aa50062f25d4fc558e5555b989587ef/rosdep/python.yaml#L5088)) when this is ready for prime time

ruffsl replied:
See: 0b27a81

mikaelarguedas reviewed (commented):
> How's it now?

Better! less unnecessary permissions and it validates the schema!

Few surprising things in the result file:

- empty profiles:
Profiles are generated for namespaces that don't belong to them and this results in a lot of empty profiles.
```
    <profile node="gazebo" ns="/global_costmap"/>
    <profile node="gazebo" ns="/local_costmap"/>
```

- There are some topics listed in the database that don't appear in the generated file, for example `tf` doesnt appear for any node in the generated policies o_O

- `clock` appears sometimes absolute and sometimes relative. I believe it makes sense as when the namespace is just `/` the script has no way to know which one of the two it is. Something to keep in mind when adding logic to collapse permissions to higher level common objects

ruffsl committed:
Fix unintended node and namespace combinatorial
node names should be limited to what exists
with respect to the current iterative namespace

Signed-off-by: ruffsl <roxfoxpox@gmail.com>

ruffsl committed:
Declare python dependency for pandas

Signed-off-by: ruffsl <roxfoxpox@gmail.com>

ruffsl committed:
Check for unintelligible or missing QoS data

Drop rows with non-parcelable userdata payloads,
as we can't infer the node name or namespace from participant

Participant vs data discovery sent over separate QoS packets in DDS.
Ensure a profile is created even if no topics are discovered.
E.g. for the ros2 cli daemon only listing nodes on the network

Signed-off-by: ruffsl <roxfoxpox@gmail.com>

ruffsl committed:
Fix typos in compression

Signed-off-by: ruffsl <roxfoxpox@gmail.com>

ruffsl commented:
> * empty profiles:
>   Profiles are generated for namespaces that don't belong to them and this results in a lot of empty profiles.

Empty profiles were unintendedly added with an erroneous combinatorial across nodes and namespaces: address in 368f4c5 . I've also added a check for when the discovery user_data from a foreign QoS configuration: see commit message in 3533ec2

> * There are some topics listed in the database that don't appear in the generated file, for example `tf` doesnt appear for any node in the generated policies o_O

The template can be debugged better commenting back the secondary transforms:

https://github.com/ros2/sros2/blob/1854b7ec50c47c0be3c2dce384b0b053c10d6e88/sros2/sros2/policy/templates/dds/demangle.xsl#L204-L215

Diffing the consecutive chained output from different iterations of commiting, I found some copy and paste typos in the compression. See 1854b7e . I'd like to make this a little more DRY, but it's tricky when only using XSL v1.0. Let me know if you see a better expression for things. 

> * `clock` appears sometimes absolute and sometimes relative. I believe it makes sense as when the namespace is just `/` the script has no way to know which one of the two it is. Something to keep in mind when adding logic to collapse permissions to higher level common objects

Yea, I thought about this, like adding a lookup for common root level topics in the xst for such cases, but it's probably better instead to ensure the added logic to collapse permissions to higher level common objects is robust enough to handle both appearances, perhaps by expanding the fqn topic namespace using rcl logic when comparing candidate substitutions?

ruffsl commented:
You can diff from last pastebin using the fix output here: https://pastebin.com/WYvxNTbg

ruffsl committed:
Fix file output

Signed-off-by: ruffsl <roxfoxpox@gmail.com>

ruffsl committed:
Add raw permission type
Update schema and transforms
as well as generating script

Signed-off-by: ruffsl <roxfoxpox@gmail.com>

mikaelarguedas commented:
Deduplicating the topics in the raw intermediate representation now generate the exact same topic list as https://github.com/mikaelarguedas/tb3_demo/blob/857e1c62ba06476cda4ba359fdf59d31ba488f91/policies/sqlite_playground/policy_dump_from_sqlite.xml#L1094 :tada: 

```diff
diff --git a/sros2/scripts/dds_sql_to_sros2_policy.py b/sros2/scripts/dds_sql_to_sros2_policy.py
index 6c602cd..801a919 100755
--- a/sros2/scripts/dds_sql_to_sros2_policy.py
+++ b/sros2/scripts/dds_sql_to_sros2_policy.py
@@ -91,9 +91,12 @@ def df_to_dds_policy(df):
                 if not pd.isna(mode):
                     topics = etree.SubElement(profile, 'raws')
                     topics.set(mode, "ALLOW")
+                    raw_topics = []
                     for dds_topic in df['dds_topic'].loc[namespace, name, mode]:
-                        topic = etree.SubElement(topics, 'raw')
-                        topic.text = dds_topic
+                        if dds_topic not in raw_topics:
+                            raw_topics.append(dds_topic)
+                            topic = etree.SubElement(topics, 'raw')
+                            topic.text = dds_topic
     return dds_policy
```

Comparing the second stage is a bit trickier but spot checking a few nodes everything seems to work as expected :+1:

>  I'd like to make this a little more DRY, but it's tricky when only using XSL v1.0. Let me know if you see a better expression for things.

Yeah that would be good indeed, but like you said it looks tricky to do with XSL 1.0. We can do more investigation after we get everything running well.

>  like adding a lookup for common root level topics in the xst for such cases, but it's probably better instead to ensure the added logic to collapse permissions to higher level common objects is robust enough to handle both appearances, perhaps by expanding the fqn topic namespace using rcl logic when comparing candidate substitutions?

Yep that's the approach I had in mind (using FQNs) before comparing to common assets.
Not sure if rcl logic needs to be leveraged here as the concatenation of namespace and names already happens in the xsl in some places.

ruffsl committed:
Separate raw changes into different schema for now

Signed-off-by: ruffsl <roxfoxpox@gmail.com>

ruffsl commented:
@mikaelarguedas and I chatted offline about adding a `<raw>` type element to the policy schema. Though I think we'll hold off on that for now until we have a more concrete use case for it. So I've just relegated that schema check to its own transport folder for now. a67dbc7
